### ISSUE TITLE
Type "(T1, T2) -> R" cannot be assigned to type "(T1, *tuple[Any, ...]) -> R": Parameter "*__p1[0]" has no corresponding parameter 

### ISSUE THREAD

**Issue by qexat** [microsoft/pyright#4551]
**Disclaimer**

`mypy` seems to have an equivalent behavior that I call "incorrect" so I'm not 100% sure that it is a bug or if it is intended -- I'm reporting it anyway just in case, maybe this (supposedly) bug is on both software.

**Describe the bug**

Here's a minimal example:
```py
class my_decorator:
    def __init__(self, func: Callable[[int, *tuple[Any, ...]], int]) -> None:
        ...

@my_decorator
def my_func(arg1: int, arg2: str, arg3: float) -> int:
    ...
```

The following error is reported on `@my_decorator`:
```
Argument of type "(arg1: int, arg2: str, arg3: float) -> int" cannot be assigned to parameter "func" of type "(int, *tuple[Any, ...]) -> int" in function "__init__"
  Type "(arg1: int, arg2: str, arg3: float) -> int" cannot be assigned to type "(int, *tuple[Any, ...]) -> int"
    Parameter "*__p1[0]" has no corresponding parameter
```

**Expected behavior**
I think that no errors should be reported.

**VS Code extension or command-line**
Running Pyright from Pylance in VS Code.
Pylance version: v2023.1.40 

**Additional context**

This piece of code:
```py
class my_decorator:
    def __init__(self, func: Callable[[int, *tuple[str, ...]], int]) -> None:
        ...

@my_decorator
def my_func(arg1: int, arg2: str, arg3: str) -> int:
    ...
```
also reports the same error ; it is not tied to `Any`.

As expected, replacing `*` by `Unpack` does not change the behavior.

**Comment by qexat** [comment]
I don't know how to properly title this issue, sorry in advance.

**Comment by qexat** [comment]
Some folk pointed out to me that unpacked items in `Callable` are treated as elements for `*args`.
But this following piece of code also triggers the error, on both Pyright and mypy...
```py
from typing import Any, Callable
from typing_extensions import Unpack


class my_decorator:
    def __init__(self, func: Callable[[int, Unpack[tuple[Any, ...]]], int]) -> None:
        ...


@my_decorator
def my_func(arg1: int, *args: Unpack[tuple[str, float]]) -> int:
    return arg1
```

Pyright error:
```
Argument of type "(arg1: int, *args: *tuple[str, float]) -> int" cannot be assigned to parameter "func" of type "(int, *tuple[Any, ...]) -> int" in function "__init__"
  Type "(arg1: int, *args: *tuple[str, float]) -> int" cannot be assigned to type "(int, *tuple[Any, ...]) -> int"
    Parameter "*__p1[0]" has no corresponding parameter
```
Mypy error:
```
t.py:10: error: Argument 1 to "my_decorator" has incompatible type "Callable[[int, VarArg(Unpack[Tuple[str, float]])], int]"; expected "Callable[[int, Unpack[Tuple[Any, ...]]], int]"  [arg-type]
```

**Comment by erictraut** [comment]
Mypy and pyright are correct here. The type `Callable[[int, *tuple[Any, ...]], int]` represents a callable that can accept any number of positional arguments. It therefore must contain a `*args` parameter, but `my_func` does not.

**Comment by qexat** [comment]
> Mypy and pyright are correct here. The type `Callable[[int, *tuple[Any, ...]], int]` represents a callable that can accept any number of positional arguments. It therefore must contain a `*args` parameter, but `my_func` does not.

I got it. That said, I still don't understand why the code in my latest comment also fails, given that it contains a `*args`...

**Comment by erictraut** [comment]
The code in your last comment doesn't accept an arbitrary number of positional arguments. It accepts precisely three: an `int`, a `str` and a `float`. If you attempted to call this function with, say, four arguments, you would receive a runtime exception.

**Comment by qexat** [comment]
Makes sense, thanks for the enlightenment.
